libflux/flux-core/src/semantic/mod.rs

```diff
@@ -228,17 +320,14 @@ impl<'env, I: import::Importer> Analyzer<'env, I> {
             }
         }
 
-        let env = self.env.clone();
-        // Clone the environment as the inferred package may mutate it.
-        let env = match nodes::infer_package(&mut sem_pkg, env, sub, &mut self.importer) {
-            Ok(env) => ExportEnvironment {
-                values: env
-                    .values
-                    .into_iter()
-                    .map(|(k, v)| (k.to_string(), v))
-                    .collect(),
-            },
+        self.env.enter_scope();
+        let env = match nodes::infer_package(&mut sem_pkg, &mut self.env, sub, &mut self.importer) {
+            Ok(()) => {
+                let env = self.env.exit_scope();
+                PackageExports::try_from(env.values)?
+            }
             Err(err) => {
+                self.env.exit_scope();
```

Comment:
I might be confused, but can the addition of `enter_scope` and `exit_scope` affect how the `flux repl` works? I am unclear on the details, but somehow it's possible for users to add assignment statements to the topmost scope incrementally so they are visible in subsequent input.

There are no unit tests for the REPL so if it broke somehow I don't think we'd catch it in CI.

Reply:
There are tests that run multiple eval calls where variables from an earlier scope are available in later ones which I assumed were the REPL ones, or at least close to it so we should be good there.

All new bindings (and only the new bindings) are now returned now and the `StatefulAnalyzer` updates its environment with those after `analyze_ast` is called which gives the same behaviour (or close enough).